components/misc_metrics/page_metrics.cc

```diff
@@ -10,59 +10,128 @@
 #include "base/functional/bind.h"
 #include "base/logging.h"
 #include "base/time/time.h"
+#include "brave/components/brave_shields/content/browser/brave_shields_util.h"
 #include "brave/components/misc_metrics/pref_names.h"
 #include "brave/components/p3a_utils/bucket.h"
 #include "brave/components/time_period_storage/weekly_storage.h"
+#include "components/content_settings/core/browser/host_content_settings_map.h"
 #include "components/history/core/browser/history_service.h"
 #include "components/prefs/pref_registry_simple.h"
 #include "components/prefs/pref_service.h"
+#include "components/security_interstitials/core/https_only_mode_metrics.h"
 
 namespace misc_metrics {
 
 namespace {
 
 constexpr const int kPagesLoadedBuckets[] = {0, 10, 50, 100, 500, 1000};
 constexpr const int kDomainsLoadedBuckets[] = {0, 4, 10, 30, 50, 100};
+constexpr const int kFailedHTTPSUpgradeBuckets[] = {0, 25, 50, 100, 300, 700};
 
-constexpr base::TimeDelta kPagesLoadedReportInterval = base::Minutes(30);
-constexpr base::TimeDelta kDomainsLoadedReportInterval = base::Minutes(30);
-constexpr base::TimeDelta kPagesLoadedInitReportDelay = base::Seconds(30);
-constexpr base::TimeDelta kDomainsLoadedInitReportDelay = base::Seconds(30);
+constexpr base::TimeDelta kReportInterval = base::Minutes(30);
+constexpr base::TimeDelta kInitReportDelay = base::Seconds(30);
 
 }  // namespace
 
+using HttpsEvent = security_interstitials::https_only_mode::Event;
+
 PageMetrics::PageMetrics(PrefService* local_state,
-                                       history::HistoryService* history_service)
-    : local_state_(local_state), history_service_(history_service) {
+                         HostContentSettingsMap* host_content_settings_map,
+                         history::HistoryService* history_service)
+    : local_state_(local_state),
+      host_content_settings_map_(host_content_settings_map),
+      history_service_(history_service) {
   DCHECK(local_state);
   DCHECK(history_service);
 
-  pages_loaded_report_timer_.Start(FROM_HERE, kPagesLoadedReportInterval, this,
-                                   &PageMetrics::ReportPagesLoaded);
-  domains_loaded_report_timer_.Start(FROM_HERE, kDomainsLoadedReportInterval,
-                                     this,
-                                     &PageMetrics::ReportDomainsLoaded);
-
-  pages_loaded_report_init_timer_.Start(FROM_HERE, kPagesLoadedInitReportDelay,
-                                        this,
-                                        &PageMetrics::ReportPagesLoaded);
-  domains_loaded_report_init_timer_.Start(
-      FROM_HERE, kDomainsLoadedInitReportDelay, this,
-      &PageMetrics::ReportDomainsLoaded);
+  init_timer_.Start(FROM_HERE, kInitReportDelay, this,
+                    &PageMetrics::ReportAllMetrics);
+
+  https_navigation_event_observer_ =
+      std::make_unique<base::StatisticsRecorder::ScopedHistogramSampleObserver>(
```

Comment:
AFAIU, page metrics instances are created one per profile, this way we would have one observer for each profile, that could record a single histogram multiple times?

Reply:
yep, this is the case with many of our profile-specific metrics. i'm not concerned since 98% of our users have one profile.